src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ClientConnectionManager.cs

```diff
@@ -0,0 +1,135 @@
+                    var errorResponse = GetMemoryString(responseStream);
+                    throw new InvalidOperationException(errorResponse);
+                }
+                return (AzureTransport)hostContext.Environment[Constants.Context.AzureSignalRTransportKey];
+            }
+
+            // This happens when hub is not found
+            Debug.Fail("Unauthorized");
+            throw new InvalidOperationException("Unable to authorize request");
+        }
+
+        internal HostContext GetHostContext(OpenConnectionMessage message, Stream responseStream)
+        {
+            var connectionId = message.ConnectionId;
+            var context = new OwinContext();
+            var response = context.Response;
+            var request = context.Request;
+
+            response.Body = responseStream;
+
+            var user = request.User = GetUserPrincipal(message);
+
+            request.Path = new PathString("/");
+
+            var userToken = string.IsNullOrEmpty(user.Identity.Name) ? string.Empty : ":" + user.Identity.Name;
+
+            var queryCollection = HttpUtility.ParseQueryString(message.QueryString ?? string.Empty);
+            queryCollection[Constants.QueryString.ConnectionToken] = $"{connectionId}{userToken}";
+
+            request.QueryString = new QueryString(queryCollection.ToString());
+
+            if (message.Headers != null)
+            {
+                foreach (var pair in message.Headers)
+                {
+                    request.Headers.Add(pair.Key, pair.Value);
+                }
+            }
+
+            context.Environment[Constants.Context.AzureServiceConnectionKey] = this;
+            return new HostContext(context.Environment);
+        }
+
+        internal static ClaimsPrincipal GetUserPrincipal(OpenConnectionMessage message)
+        {
+            if (message.Claims == null || message.Claims.Length == 0)
+            {
+                return EmptyPrincipal;
+            }
+
+            var claims = new List<Claim>();
+            var authenticationType = "Bearer";
+
+            foreach (var claim in message.Claims)
+            {
+                // TODO: Add prefix "azure.signalr.user." to user claims instead of guessing them?
+                if (claim.Type == Constants.ClaimType.AuthenticationType)
+                {
+                    authenticationType = claim.Value;
+                }
+                else if (!SystemClaims.Contains(claim.Type) && !claim.Type.StartsWith(Constants.ClaimType.AzureSignalRSysPrefix))
+                {
+                    claims.Add(claim);
+                }
+            }
+
+            return new ClaimsPrincipal(new ClaimsIdentity(claims, authenticationType));
+        }
+
+        internal static string GetMemoryString(MemoryStream stream)
+        {
+            stream.Seek(0, SeekOrigin.Begin);
+            using (var reader = new StreamReader(stream))
```

Comment:
using `StreamReader` will close the stream, but from the method name, I haven't realized it.

Reply:
how about `GetContentAndDispose`?